### DISCUSSION TITLE
Write a multi-index pandas dataframe as a table

### DISCUSSION THREAD

**Discussion by stevenlis** [py-pdf/fpdf2#969] [Q&A]
```python
import seaborn as sns
df = sns.load_dataset('tips')
(
    df.groupby(['sex', 'smoker', 'time'])[['tip', 'size', 'total_bill']]
    .mean().round(1).reset_index()
    .pivot(index=['smoker', 'time'], columns=['sex'])
)
```
It looks like this in a Jupyter notebook environment:
![CS-xkiXnYlQ@2x](https://github.com/py-pdf/fpdf2/assets/33796896/0f391041-98c8-40e0-9607-c2f56372a76c)

I would like to write a table like this in a PDF:
![CS-wiY9oyxP@2x](https://github.com/py-pdf/fpdf2/assets/33796896/b61a7064-f693-49f1-b377-45fdc1f0fdd5)
(Please disregard the double/cell borders above. I am unsure how to control them in HTML. Single borders work fine for me)

I found an example [here](https://py-pdf.github.io/fpdf2/Maths.html), but it doesn't work on a multi-index dataframe, which is common when generating summary statistics. I'm wondering if someone could help me with this.

Thanks in advance.

**Comment by gmischler** [comment]
Give them tables, and they will want singing, dancing, and jingling tables... 🌟🔔👯🎵🎆 

The most banal obstacle to this is that our tables currently don't support rowspan.
Other than that, it seems largely to be a question of appropriate cell formatting.
The ability to have several heading lines has just been added with the latest release. Your example suggests that adding a vertical version of that (key columns?) might also be helpful. Coming up with an API that can combine the two (and their different formatting rules) without too much confusion will be an interesting challenge.

As usual in a situation like this: PRs welcome! 😉

**Comment by stevenlis** [comment]
A workaround:
```python
import seaborn as sns
df = sns.load_dataset('tips')
df = (
    df.groupby(['sex', 'smoker', 'time'])[['tip', 'size', 'total_bill']]
    .mean().round(1).reset_index()
    .pivot(index=['smoker', 'time'], columns=['sex'])
).reset_index().astype('str')

levels = df.columns.nlevels
df.columns.names = [None]*levels

columns = [df.columns.get_level_values(l).tolist() for l in range(levels)]
rows = df.values.tolist()
data = columns + rows

with pdf.table(
    line_height=pdf.font_size * 2,
    text_align='RIGHT',
    width=160
) as table:
    for data_row in data:
        row = table.row()
        for datum in data_row:
            row.cell(datum)
```
![CS-G7j1RVsM@2x](https://github.com/py-pdf/fpdf2/assets/33796896/a3369f33-8c7a-4dda-b8ae-8a4b745c2ffa)

**Comment by Lucas-C** [comment]
For the record, there has been a follow-up of this discussion in https://github.com/py-pdf/fpdf2/issues/970

**Comment by stevenlis** [comment]
Another workaround is using `to_html` and `write_html`:

```python
from fpdf import FPDF
import seaborn as sns

pdf = FPDF()
pdf.set_font_size(16)
pdf.add_page()

df = sns.load_dataset('tips')
string_html = (
    df.groupby(['sex', 'smoker', 'time'])[['tip', 'size', 'total_bill']]
    .mean().round(1).reset_index()
    .pivot(index=['smoker', 'time'], columns=['sex'])
).reset_index().to_html()

pdf.write_html(string_html)
pdf.output('table-pdf.pdf')
```

![CS-kbCfiLIf@2x](https://github.com/py-pdf/fpdf2/assets/33796896/58c9b3ed-1c3f-4950-b4d1-498517280c10)
